codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java

```diff
@@ -104,50 +105,60 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
-                        WebSocketExtensionData extensionData = extension.newReponseData();
-                        headerValue = WebSocketExtensionUtil.appendExtension(headerValue,
-                                                                             extensionData.name(),
-                                                                             extensionData.parameters());
-                    }
-                    promise.addListener(new ChannelFutureListener() {
-                        @Override
-                        public void operationComplete(ChannelFuture future) {
-                            if (future.isSuccess()) {
-                                for (WebSocketServerExtension extension : validExtensions) {
-                                    WebSocketExtensionDecoder decoder = extension.newExtensionDecoder();
-                                    WebSocketExtensionEncoder encoder = extension.newExtensionEncoder();
-                                    ctx.pipeline()
-                                       .addAfter(ctx.name(), decoder.getClass().getName(), decoder)
-                                       .addAfter(ctx.name(), encoder.getClass().getName(), encoder);
-                                }
-                            }
-                        }
-                    });
+        if (WebSocketExtensionUtil.isWebsocketUpgrade(headers)) {
 
-                    if (headerValue != null) {
-                        headers.set(HttpHeaderNames.SEC_WEBSOCKET_EXTENSIONS, headerValue);
-                    }
-                }
+            if (validExtensions != null) {
+                String headerValue = headers.getAsString(HttpHeaderNames.SEC_WEBSOCKET_EXTENSIONS);
 
+                for (WebSocketServerExtension extension : validExtensions) {
+                    WebSocketExtensionData extensionData = extension.newReponseData();
+                    headerValue = WebSocketExtensionUtil.appendExtension(headerValue,
+                        extensionData.name(),
+                        extensionData.parameters());
+                }
                 promise.addListener(new ChannelFutureListener() {
                     @Override
                     public void operationComplete(ChannelFuture future) {
                         if (future.isSuccess()) {
-                            ctx.pipeline().remove(WebSocketServerExtensionHandler.this);
+                            for (WebSocketServerExtension extension : validExtensions) {
+                                WebSocketExtensionDecoder decoder = extension.newExtensionDecoder();
+                                WebSocketExtensionEncoder encoder = extension.newExtensionEncoder();
+                                ctx.pipeline()
+                                    .addAfter(ctx.name(), decoder.getClass().getName(), decoder)
+                                    .addAfter(ctx.name(), encoder.getClass().getName(), encoder);
```

Comment:
As we're already here, we can move **ctx.name()** to own local variables. That will help the compiler a bit.

Reply:
Good idea, especially if this code isn't going to get hot most of the time